### PR TITLE
feat(gen) add Dockerfile and docker-compose.yml support

### PR DESCRIPTION
**Pull request by kennu**
Add support to run generated Angular-Fullstack applications as Docker containers by including
Dockerfile, docker-compose.yml and .dockerignore. When the application is built with grunt
build, these files are copied to the dist directory. The Docker container can then be built
and started there with the "docker-compose up" command. When Mongoose is used, Docker Compose
will also launch a MongoDB container, eliminating the need to install MongoDB on the local
machine. Note that Docker and Docker Compose must be installed for Docker support to work.
They are available on Linux, OS X and Windows.

### CODE REVIEW

**Comment by Awk34** [comment]
@kennu have time to re-visit this?

**Comment by kingcody** [comment]
It would be nice if we could deploy a docker image from a development repo. For instance you could use an automated build on the docker hub and point it to your project's git repo. However we would need a more complex dockerfile like the one I posted [here](https://github.com/DaftMonk/generator-angular-fullstack/issues/516#issuecomment-55516168). The difference is that setup builds the entire project in the docker image and then sets it up for production use.

Pros:
* Consistent build environment
* Build directly from the source repo

Cons:
* Larger docker images (will contain build deps)


**`+ 2¢`**

**Comment by kennu** [comment]
Hi guys, I rebased the changes but still need to fix something to pass the tests.

**Comment by kennu** [comment]
Now the tests pass, but for some reason I was unable to add a new test that would specify docker: true and generate Dockerfile and other files. The generator just ignored the setting.

File: app/templates/Dockerfile(docker) [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Production Dockerfile - Used in the dist directory after grunt build
+FROM node:0.12-onbuild
```
Comment (Awk34) [inline app/templates/Dockerfile(docker)]: We shouldn't be using Node < 4

File: app/templates/server/config/environment/production.js [hunk 2]
```diff
@@ -18,7 +18,11 @@ module.exports = {
     uri:    process.env.MONGOLAB_URI ||
             process.env.MONGOHQ_URL ||
             process.env.OPENSHIFT_MONGODB_DB_URL +
-            process.env.OPENSHIFT_APP_NAME ||
+            process.env.OPENSHIFT_APP_NAME ||<% if (filters.docker) { %>
+            (process.env.MONGODB_PORT_27017_TCP_ADDR ? 'mongodb://' +
+              process.env.MONGODB_PORT_27017_TCP_ADDR + ':' +
+              process.env.MONGODB_PORT_27017_TCP_PORT +
+              '/<%= _.slugify(appname) %>' : null) ||<% } %>
```
Comment (Awk34) [inline app/templates/server/config/environment/production.js]: What's all this?

**Comment by kennu** [comment]
Sorry, I'm not really using Docker any more for this kind of projects so I can't put much effort into this. (The original pull request was made over a year ago.)

**Comment by Awk34** [comment]
@kennu yeah I feared that, this PR has been open for a long time.

File: app/templates/server/config/environment/production.js (hunk 2 above)
Comment (JeremyMarshall) [inline app/templates/server/config/environment/production.js]: This is code to make the docker node container see the mongo container https://www.thachmai.info/2015/05/10/docker-container-linking-mongo-node/

**Pull request closed without merging**